sql: flake in TestDropTableDeleteData

Seen on unrelated change under race [here](https://teamcity.cockroachdb.com/viewLog.html?buildId=12040252&buildTypeId=Cockroach_BazelExtendedCi):
```
Failed
=== RUN   TestDropTableDeleteData
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/b287de7194e17a0386ee1775f0d198ba/logTestDropTableDeleteData2309674462
    test_log_scope.go:81: use -show-logs to present logs inline
    drop_test.go:724: descriptor exists = true
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/b287de7194e17a0386ee1775f0d198ba/logTestDropTableDeleteData2309674462
--- FAIL: TestDropTableDeleteData (143.66s)
```

Jira issue: CRDB-32046

--- a/pkg/sql/drop_test.go
+++ b/pkg/sql/drop_test.go
@@ -676,6 +676,14 @@ func TestDropTableDeleteData(t *testing.T) {
 	// Speed up mvcc queue scan.
 	params.ScanMaxIdleTime = time.Millisecond
 
+	allowGC := make(chan struct{})
+	params.Knobs.GCJob = &sql.GCJobTestingKnobs{
+		RunBeforePerformGC: func(jobID jobspb.JobID) error {
+			<-allowGC
+			return nil
+		},
+	}
+
 	s, sqlDB, kvDB := serverutils.StartServer(t, params)
 	defer s.Stopper().Stop(context.Background())
 
@@ -736,7 +744,12 @@ func TestDropTableDeleteData(t *testing.T) {
 		}
 		tableSpan := descs[i].TableSpan(keys.SystemSQLCodec)
 		tests.CheckKeyCountIncludingTombstoned(t, s, tableSpan, numKeys)
+	}
 
+	close(allowGC)
+
+	// Now verify that the GC job has run.
+	for i := 0; i < numTables; i++ {
 		if err := jobutils.VerifySystemJob(t, sqlRun, numTables+i,
 			jobspb.TypeNewSchemaChange, jobs.StatusSucceeded, jobs.Record{
 				Username:    username.RootUserName(),

